2655 patching bad fit

DanielNobbe opened the pull request:
#### Changes
<!-- Please include a summary of the change and which issue is fixed. -->
<!-- Please also include relevant motivation and context. -->
<!-- List any dependencies that are required for this change. -->

Fixes #2655: By default, the `folding` method used inside `extract_tensor_patches` drops any remainders, i.e. any parts of the image that would result in a partial patch. This has the additional downside that if no padding was specified, the `combine_tensor_patches` function could no longer reconstruct the image, since the `original_size` is no longer correct.   
This PR fixes this by adding an option to automatically add patching to `extract_tensor_patches`, and automatically remove padding through the same calculations in `combine_tensor_patches`.    
Calculating if the current image fits into the stride and window size uses the assumption that there needs to be an integer amount of strides within the image width/height minus one window size (half a window size to the 'left' of the first patch centre, and half a window size to the 'right' of the last patch centre). 


#### Type of change
<!-- Please delete options that are not relevant. -->
- [ ] 📚  Documentation Update
- [x] 🧪 Tests Cases
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🔬 New feature (non-breaking change which adds functionality)
- [ ] 📝 This change requires a documentation update


#### Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

edgarriba reviewed (commented):
consider  https://github.com/kornia/kornia/pull/2650

DanielNobbe commented:
> consider #2650

Thanks for pointing me to that PR. Don't think it resolves my issue, but I'll make sure this PR is compatible with the changes there

johnnv1 commented on kornia/contrib/extract_patches.py:
```diff
@@ -249,6 +251,40 @@ def combine_tensor_patches(
             "Please feel free to drop a PR to Kornia Github."
         )
 
+    # now check if we need to do automatic unpadding
```
can you merge it within the master? The #2650 have been merged

johnnv1 reviewed (commented):
needs tests too -- look at contributing as a base: https://github.com/kornia/kornia/blob/master/CONTRIBUTING.md#coding-standards

edgarriba commented on kornia/contrib/extract_patches.py:
```diff
@@ -8,22 +8,44 @@
 
 from kornia.core import Module, Tensor, pad
 
+FullPadType = Tuple[int, int, int, int]
```
```suggestion
FullPadType = tuple[int, int, int, int]
```
can we use `from __future__ import annotations` ?

DanielNobbe replied:
This apparently doesn't work for a TypeAlias in 3.8, I'm not sure why (so this line doesn't get accepted by mypy). I'd rather keep everything with `Tuple` rather than some things `Tuple` and some `tuple`.

johnnv1 commented on kornia/contrib/extract_patches.py:
```diff
@@ -9,11 +9,12 @@
 
 def create_padding_tuple(padding: PadType, unpadding: bool = False) -> FullPadType:
-    padding = _pair(padding)
+    padding = cast(TuplePadType, _pair(padding))
```
Should we expose the https://github.com/kornia/kornia/blob/619d65f9caae73016cad33d7aa006be4095a1164/kornia/filters/kernels.py#L26 on kornia core or something? this way we don't need to use this internal ops from torch. @edgarriba

edgarriba commented:
@DanielNobbe @johnnv1  we are releasing by the end of the week. From my side this PR sounds enough. Do you think we can make it happen before that ?

DanielNobbe commented:
> @DanielNobbe @johnnv1 we are releasing by the end of the week. From my side this PR sounds enough. Do you think we can make it happen before that ?

I just did some final tests, and found a small bug in my padding algorithm when the stride fit into the final patch (it would attempt to add another patch even if the centre of the true final patch was still inside the image boundaries). I fixed it in my last commits. Another bug I found prevented the auto padding from being used when using the ExtractTensorPatches and CombineTensorPatches classes, which I also fixed.
In addition, I added support for patching with stride > window size, which is now only when combining patches.

Most tests pass on my side, but I see there are a few CI tests failing. I'm not sure what these entail though, could you shine a light there?

One thing that still needs to be fixed is the tutorial notebook -- the new padding algorithm I implemented is a bit more efficient with its padding, so there are fewer patches in the example image in the `extract_combine_patches.ipynb` notebook. As a result, the plotting looks a bit off. Are the tutorials on the same release schedule? If so I can try to fit it in this week.

johnnv1 commented:
The failing test looks like something isn't generating the patches at all when testing the module, because all these generates with patch dim equals one

johnnv1 commented on kornia/contrib/extract_patches.py:
```diff
@@ -32,22 +59,44 @@ def compute_padding(
+        right_padding = ceil(horizontal_padding / 2)
+    # the new implementation with unfolding requires symmetric padding
+    padding = int(top_padding), int(bottom_padding), int(left_padding), int(right_padding)
+    return cast(FullPadType, padding)
```
this cast shouldn't be required

DanielNobbe commented:
> The failing test looks like something isn't generating the patches at all when testing the module, because all these generates with patch dim equals one

edgarriba commented:
@DanielNobbe thanks for the effort. I think we can attempt to merge this. Ideally, we need to land this in the `main` branch (notice about the recent changes: https://github.com/kornia/kornia/discussions/2719

I verified that there are some conflicts which i think between myself and @johnnv1 we can help to rebase and resolve.

johnnv1 commented:
I squashed these commits, them cherry-pick it to put on top of the main, please check #2735
